File: workbench/min0_core_forth_publish.py
"""Safe-runtime W^X image publication for MIN0 CORE FORTH (R0)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping

from min0_core_forth_dictionary import RuntimeDictionary
from min0_core_forth_image import (
    EXECUTION_PROFILE_SAFE_RUNTIME,
    ImageError,
    validate_image_envelope,
)
from min0_core_forth_outer import OuterInterpreter
from min0_core_forth_vm import MemoryRegion, Min0CoreForthVM, RegionMemory


class PublishError(ImageError):
    pass


@dataclass(frozen=True)
class PublishedRuntime:
    vm: Min0CoreForthVM
    dictionary: RuntimeDictionary
    outer: OuterInterpreter
    staging_memory: RegionMemory
    validation: dict


def _r0_layout(bases: Mapping[str, int], limits: Mapping[str, int]) -> int:
    if not (
        0 < bases["code"] < limits["code"] == bases["dictionary"]
        < limits["dictionary"] == bases["data"] < limits["data"]
    ):
        raise PublishError(
            "R0 publication requires contiguous CODE, DICTIONARY, and DATA limits"
        )
    return limits["data"]


def publish_runtime_image(
    components: Mapping[str, object],
    envelope: object,
    *,
    require_authentication: bool = False,
    minimum_generation: int = 0,
    trusted_public_keys: Mapping[str, object] | None = None,
    required_image_role: str | None = None,
) -> PublishedRuntime:
    """Validate in non-executable staging, copy to runtime, then seal CODE rx."""

    validation_options = {
        "require_authentication": require_authentication,
        "minimum_generation": minimum_generation,
        "trusted_public_keys": trusted_public_keys,
        "required_image_role": required_image_role,
        "runtime_profile": EXECUTION_PROFILE_SAFE_RUNTIME,
    }
    first = validate_image_envelope(components, envelope, **validation_options)
    bases = first["bases"]
    limits = first["limits"]
    memory_size = _r0_layout(bases, limits)

    staging = RegionMemory(
        memory_size,
        [
            MemoryRegion(
                "STAGING-CODE", bases["code"],
                limits["code"] - bases["code"], "rw",
            ),
            MemoryRegion(
                "STAGING-DICTIONARY", bases["dictionary"],
                limits["dictionary"] - bases["dictionary"], "rw",
            ),
            MemoryRegion(
                "STAGING-DATA", bases["data"],
                limits["data"] - bases["data"], "rw",
            ),
        ],
    )
    for section in ("code", "dictionary", "data"):
        staging.write(bases[section], first["components"][section])
    staged_components = {
        section: staging.read(bases[section], len(first["components"][section]))
        for section in ("code", "dictionary", "data")
    }
    validated = validate_image_envelope(
        staged_components, envelope, **validation_options
    )

    code_region = MemoryRegion(
        "CODE", 0, limits["code"], "rx", programmable=True
    )
    runtime_memory = RegionMemory(
        memory_size,
        [
            code_region,
            MemoryRegion(
                "DICTIONARY", bases["dictionary"],
                limits["dictionary"] - bases["dictionary"], "rw",
            ),
            MemoryRegion(
                "DATA", bases["data"],
                limits["data"] - bases["data"], "rw",
            ),
        ],
    )
    vm = Min0CoreForthVM(memory_size=memory_size, memory_bus=runtime_memory)
    vm.load(staged_components["code"], bases["code"])
    dictionary = RuntimeDictionary(
        vm,
        base=bases["dictionary"],
        limit=limits["dictionary"],
        body_base=bases["data"],
        body_limit=limits["data"],
    )
    allocator = validated["allocator"]
    dictionary.load_images(
        staged_components["dictionary"],
        latest=allocator["latest"],
        body_image=staged_components["data"],
    )
    outer = OuterInterpreter(vm, dictionary, code_base=allocator["code_here"])
    dictionary.seal_runtime_structure()
    vm.seal_verified_execution(
        validated["code_verification"],
        extra_entries=outer.execution_extra_entries(),
    )
    return PublishedRuntime(vm, dictionary, outer, staging, validated)

File: workbench/min0_core_forth_publish.rb
# frozen_string_literal: true

require_relative "min0_core_forth_image"
require_relative "min0_core_forth_outer"

module Min0CoreForth
  class PublishError < ImageError; end

  PublishedRuntime = Data.define(:vm, :dictionary, :outer, :staging_memory, :validation)

  module RuntimePublisher
    module_function

    def publish(
      components, envelope, require_authentication: false, minimum_generation: 0,
      trusted_public_keys: nil, required_image_role: nil
    )
      validation_options = {
        require_authentication: require_authentication,
        minimum_generation: minimum_generation,
        trusted_public_keys: trusted_public_keys,
        required_image_role: required_image_role,
        runtime_profile: EXECUTION_PROFILE_SAFE_RUNTIME
      }
      first = ImageEnvelope.validate(components, envelope, **validation_options)
      bases = first[:bases]
      limits = first[:limits]
      memory_size = r0_layout(bases, limits)

      staging = RegionMemory.new(
        memory_size,
        [
          MemoryRegion.new(
            name: "STAGING-CODE", start: bases["code"],
            size: limits["code"] - bases["code"], permissions: "rw"
          ),
          MemoryRegion.new(
            name: "STAGING-DICTIONARY", start: bases["dictionary"],
            size: limits["dictionary"] - bases["dictionary"], permissions: "rw"
          ),
          MemoryRegion.new(
            name: "STAGING-DATA", start: bases["data"],
            size: limits["data"] - bases["data"], permissions: "rw"
          )
        ]
      )
      %w[code dictionary data].each do |section|
        staging.write(bases.fetch(section), first[:components].fetch(section))
      end
      staged_components = %w[code dictionary data].to_h do |section|
        image = first[:components].fetch(section)
        [section, staging.read(bases.fetch(section), image.bytesize)]
      end
      validated = ImageEnvelope.validate(
        staged_components, envelope, **validation_options
      )

      code_region = MemoryRegion.new(
        name: "CODE", start: 0, size: limits["code"],
        permissions: "rx", programmable: true
      )
      runtime_memory = RegionMemory.new(
        memory_size,
        [
          code_region,
          MemoryRegion.new(
            name: "DICTIONARY", start: bases["dictionary"],
            size: limits["dictionary"] - bases["dictionary"], permissions: "rw"
          ),
          MemoryRegion.new(
            name: "DATA", start: bases["data"],
            size: limits["data"] - bases["data"], permissions: "rw"
          )
        ]
      )
      vm = VM.new(memory_size: memory_size, memory_bus: runtime_memory)
      vm.load(staged_components.fetch("code"), address: bases.fetch("code"))
      dictionary = RuntimeDictionary.new(
        vm,
        base: bases.fetch("dictionary"), limit: limits.fetch("dictionary"),
        body_base: bases.fetch("data"), body_limit: limits.fetch("data")
      )
      allocator = validated[:allocator]
      dictionary.load_images(
        staged_components.fetch("dictionary"),
        latest: allocator.fetch("latest"),
        body_image: staged_components.fetch("data")
      )
      outer = OuterInterpreter.new(
        vm, dictionary, code_base: allocator.fetch("code_here")
      )
      dictionary.seal_runtime_structure
      vm.seal_verified_execution(
        validated[:code_verification], extra_entries: outer.execution_extra_entries
      )
      PublishedRuntime.new(vm, dictionary, outer, staging, validated)
    end

    def r0_layout(bases, limits)
      valid = bases["code"].positive? && bases["code"] < limits["code"] &&
              limits["code"] == bases["dictionary"] &&
              bases["dictionary"] < limits["dictionary"] &&
              limits["dictionary"] == bases["data"] &&
              bases["data"] < limits["data"]
      unless valid
        raise PublishError,
              "R0 publication requires contiguous CODE, DICTIONARY, and DATA limits"
      end
      limits["data"]
    end
    private_class_method :r0_layout
  end
end
